### PR TITLE
Wave 1

### PR DESCRIPTION
**Pull request by laerreal**
Next wave of patches from `devel`.

# v2
- add new patches on top to fix `SyntaxWarning` under Py3.8
- explicitly forbid user to pass window and anchor to `CanvasFrame`
  (a new patch, not squashed)
- fixed typos in commit messages and comments
- squashed 'init_submodules_from_cache: update reverted URL in submodule cache "config"'
- dot in 'QemuVersionCache.co_propagate_new_param: clarify docstring'

### CODE REVIEW

File: widgets/canvas_frame.py [hunk 1]
```diff
@@ -93,7 +94,11 @@ def __init__(self, canvas, x, y, *a, **kw):
         self.x, self.y = RESIZE_GAP * 2, RESIZE_GAP * 2
         self.__cursor = self.cget("cursor")
 
-        self.id = canvas.create_window(x, y, window = self, anchor = NW)
+        canvas_kw = dict(canvas_kw) # preserve user's `dict`
+        canvas_kw["window"] = self
+        canvas_kw["anchor"] = NW
```
Comment (BDanAnd) [inline widgets/canvas_frame.py]: Maybe the user's choice is more important?
```
canvas_kw.setdefault("window", self)
canvas_kw.setdefault("anchor", NW)
```

File: common/git_tools.py [hunk 2]
```diff
@@ -296,3 +333,26 @@ def git_find_commit(repo, version):
                     return ref.commit
 
         raise
+
+
+# Baed on: https://stackoverflow.com/a/25961128/7623015
```
Comment (BDanAnd) [inline common/git_tools.py]: Based

File: qemu/version_description.py [hunk 3]
```diff
@@ -721,6 +728,10 @@ def co_init_cache(self):
             # set Qemu version heuristics according to current version
             initialize_version(self.qvc.version_desc)
 
+            # Save cache after long running operation because consequent
+            # initialization can be buggy.
+            yield self.co_overwrite_cache()
```
Comment (BDanAnd) [inline qemu/version_description.py:733]: Maybe add:
```
if is_outdated:
    yield self.co_overwrite_cache()
```
and delete `is_outdated or` from code below?

File: common/pygen.py [hunk 4]
```diff
@@ -199,23 +199,27 @@ def gen_const(c):
                     normalized += "\\x%02x" % code
                 elif code == 92: # \
                     normalized += "\\\\"
+                # Do not use multiline strings ("""/''') because of auto
+                # indentation.
```
Comment (BDanAnd) [inline common/pygen.py:203]: `multiline` variable has become unused.

File: qemu/qom_hierarchy.py [hunk 5]
```diff
@@ -94,6 +91,35 @@ def find(self, **request):
...
+
+            if n in children:
+                c = children[n]
+            else:
+                c = type(other_c)(
+                    name = n,
+                    macros = other_c.macros,
+                    arches = other_c.arches,
+                )
+                add_child(c)
```
Comment (BDanAnd) [inline qemu/qom_hierarchy.py:116]: Is this much faster than specifying a parent?
```
                c = type(other_c)(
                    name = n,
                    parent = self,
                    macros = other_c.macros,
                    arches = other_c.arches,
                )

**Review by BDanAnd — COMMENTED** [review]
v1 viewed

File: widgets/canvas_frame.py (hunk 1 above)
Comment (laerreal) [reply widgets/canvas_frame.py]: Exactly `CanvasFrame` (`self`) is wrapped in the canvas. So, actually, user forbidden to pass `window`. Looks like anchor is not selected randomly too...

Comment (laerreal) [reply widgets/canvas_frame.py]: upd, different anchor breaks resizing

File: qemu/qom_hierarchy.py (hunk 5 above)
Comment (laerreal) [reply qemu/qom_hierarchy.py:116]: Don't know. U may try it on devel.

File: common/pygen.py (hunk 4 above)
Comment (laerreal) [reply common/pygen.py:203]: Looks like a patch on top of devel.

File: qemu/version_description.py (hunk 3 above)
Comment (laerreal) [reply qemu/version_description.py:733]: Looks like a patch on top of devel.

Comment (laerreal) [reply qemu/version_description.py:733]: (try it, and it could be included in a next wave)

File: qemu/qom_hierarchy.py (hunk 5 above)
Comment (laerreal) [reply qemu/qom_hierarchy.py:116]: (...same...)

**Comment by laerreal** [comment]
v2 is ready

**Pull request merged**